Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,156 @@
+    -- lowest common multiple of previousPoly and candidateMinPoly
+    potentialMinPoly = lcm previousPoly candidateMinPoly
+    candidateMinPoly = findCandidatePoly matrix z x
+    -- z has to be picked outside of the image of the matrix.
+    -- This works only because we know that the last row of the matrix is zero.
+    -- To solve a general singular square matrix, this won't work.
+    -- To deal with general case, it is desirable to wirte both variants.
+    -- z = DBVector $ SU.update (SU.replicate (Bit False)) $ SU.singleton (size matrix - 1, Bit True)
+
+findSolution :: KnownNat k => [Bit] -> SBMatrix k -> DBVector k -> DBVector k
+-- It is ideal to define a family of parameters indexed by the length of the vector.
+-- This is then an abelian group. In this workaround, @DBVector (U.empty)@ can
+-- be used as it is not summed to any other element.
+findSolution [] _ _ = mempty
+findSolution (_ : xs) matrix vector
+  | result == mempty = vector
+  | otherwise        = findSolution xs matrix result
+  where
+    result = matrix `mult` vector
+
+-- This routine takes a polynomial p, a matrix A and a vector w and
+-- returns p(A)w. It assumes the first coefficient of p is non zero,
+-- in particular that it is non empty. This makes the implementation
+-- easier as there is no need to write the identity matrix.
+evaluate :: KnownNat k => SBMatrix k -> DBVector k -> [Bit] -> DBVector k
+-- Here @DBVector (U.empty)@ cannot be used since summing it with any other
+-- vector gives the empty vector again.
+evaluate matrix w = foldr (\coeff acc -> (matrix `mult` acc) <> (if unBit coeff then w else mempty)) mempty
+
+findCandidatePoly :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+findCandidatePoly matrix z x = berlekampMassey dim errorPoly randomSequence
+  where
+    randomSequence = generateData matrix z x
+    errorPoly = fromInteger (1 `shiftL` (2 * dim)) :: F2Poly
+    dim = size matrix
+
+berlekampMassey :: Int -> F2Poly -> F2Poly -> F2Poly
+berlekampMassey dim = go 1 0
+  where
+    -- Is there a better way to implement recursion in this situation?
+    go :: F2Poly -> F2Poly -> F2Poly ->F2Poly -> F2Poly
+    go oneBefore twoBefore a b
+      | U.length (unF2Poly b) <= dim = oneBefore
+      -- Updated value is given by @twoBefore - oneBefore * q@
+      | otherwise                    = go (twoBefore - oneBefore * q) oneBefore b r
+        where
+          (q, r) = quotRem a b
+
+-- The input is a matrix B and a random vector z
+generateData :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+generateData matrix z x = toF2Poly $ U.fromList $ reverse $ map (x `dot`) matrixPowers
+  where
+    matrixPowers = L.take (2 * size matrix) $ L.iterate (matrix `mult`) z
+
+-- Infinite lists of random DBVectors.
+getRandomDBVectors :: KnownNat k => Int -> Double -> StdGen -> [DBVector k]
+getRandomDBVectors numberOfColumns density gen = go $ randomRs (0, 1) gen
```

Comment:
It seems redundant to pass `numberOfColumns` as a separate argument, because you already have it as `k` on the type level. One can use `natVal` to demote it as a term.